import grpc
import hashlib
import proto.user_pb2 as user_pb2
import proto.user_pb2_grpc as user_pb2_grpc

SERVER_ADDRESS = "localhost:50051"  # Change to 50051 if not using gRPC-Web Proxy

# ‚úÖ Function to Hash Password Before Sending (Security)
def hash_password(password):
    salt = b'somesalt'  # Use a random salt in production
    return hashlib.pbkdf2_hmac('sha256', password.encode(), salt, 100000)

# ‚úÖ Function to Register a User
def register_user(email_or_phone, password, username, use_biometric):
    with grpc.insecure_channel(SERVER_ADDRESS) as channel:
        stub = user_pb2_grpc.UserServiceStub(channel)
        hashed_password = hash_password(password)
        
        request = user_pb2.RegisterRequest(
            email_or_phone=email_or_phone,
            password=hashed_password.hex(),
            username=username,
            use_biometric=use_biometric
        )

        try:
            response = stub.Register(request)
            print(f"‚úÖ Registration Response: {response.message}")
        except grpc.RpcError as e:
            print(f"‚ùå Registration Failed: {e.details()}")

# ‚úÖ Function to Login a User
def login_user(email_or_phone, password):
    with grpc.insecure_channel(SERVER_ADDRESS) as channel:
        stub = user_pb2_grpc.UserServiceStub(channel)
        hashed_password = hash_password(password)

        request = user_pb2.LoginRequest(
            email_or_phone=email_or_phone,
            password=hashed_password.hex(),
            use_biometric=False  # Change this based on user preference
        )

        try:
            response = stub.Login(request)
            if response.success:
                print(f"‚úÖ Login Successful! Welcome, {response.username}")
            else:
                print(f"‚ùå Login Failed: {response.message}")
        except grpc.RpcError as e:
            print(f"‚ùå Login Failed: {e.details()}")

# ‚úÖ Interactive CLI for Testing
if __name__ == "__main__":
    print("\nüì° Connecting to gRPC Server...\n")
    while True:
        print("\n1Ô∏è‚É£ Register\n2Ô∏è‚É£ Login\n3Ô∏è‚É£ Exit")
        choice = input("Select an option: ")

        if choice == "1":
            email_or_phone = input("Enter Email or Phone: ")
            password = input("Enter Password: ")
            username = input("Enter Username: ")
            use_biometric = input("Use Biometric Login? (yes/no): ").strip().lower() == "yes"
            register_user(email_or_phone, password, username, use_biometric)

        elif choice == "2":
            email_or_phone = input("Enter Email or Phone: ")
            password = input("Enter Password: ")
            login_user(email_or_phone, password)

        elif choice == "3":
            print("üöÄ Exiting...")
            break

        else:
            print("‚ùå Invalid choice! Try again.")
